performjoin.go

```diff
@@ -0,0 +1,276 @@
+	// the make_join response.
+	// "If not provided, the room version is assumed to be either "1" or "2"."
+	// https://matrix.org/docs/spec/server_server/unstable#get-matrix-federation-v1-make-join-roomid-userid
+	roomVersion := respMakeJoin.GetRoomVersion()
+	if roomVersion == "" {
+		roomVersion = setDefaultRoomVersionFromJoinEvent(joinEvent)
+	}
+	verImpl, err := GetRoomVersion(roomVersion)
+	if err != nil {
+		return nil, &FederationError{
+			ServerName: input.ServerName,
+			Transient:  false,
+			Reachable:  true,
+			Err:        err,
+		}
+	}
+
+	// Build the join event.
+	event, err := joinEvent.Build(
+		time.Now(),
+		origin,
+		input.KeyID,
+		input.PrivateKey,
+		roomVersion,
+	)
+	if err != nil {
+		return nil, &FederationError{
+			ServerName: input.ServerName,
+			Transient:  false,
+			Reachable:  true,
+			Err:        fmt.Errorf("respMakeJoin.JoinEvent.Build: %w", err),
+		}
+	}
+
+	var respState StateResponse
+	// Try to perform a send_join using the newly built event.
+	respSendJoin, err := fedClient.SendJoin(
+		context.Background(),
+		origin,
+		input.ServerName,
+		event,
+	)
+	if err != nil {
+		return nil, &FederationError{
+			ServerName: input.ServerName,
+			Transient:  true,
+			Reachable:  false,
+			Err:        fmt.Errorf("r.federation.SendJoin: %w", err),
+		}
+	}
+
+	// If the remote server returned an event in the "event" key of
+	// the send_join response then we should use that instead. It may
+	// contain signatures that we don't know about.
+	if len(respSendJoin.GetJoinEvent()) > 0 {
+		var remoteEvent *Event
+		remoteEvent, err = verImpl.NewEventFromUntrustedJSON(respSendJoin.GetJoinEvent())
+		if err == nil && isWellFormedJoinMemberEvent(
+			remoteEvent, input.RoomID, input.UserID,
+		) {
+			event = remoteEvent
+		}
+	}
+
+	// Sanity-check the join response to ensure that it has a create
+	// event, that the room version is known, etc.
+	authEvents := respSendJoin.GetAuthEvents().UntrustedEvents(roomVersion)
+	if err = checkEventsContainCreateEvent(authEvents); err != nil {
+		return nil, &FederationError{
+			ServerName: input.ServerName,
+			Transient:  false,
+			Reachable:  true,
+			Err:        fmt.Errorf("sanityCheckAuthChain: %w", err),
+		}
+	}
+
+	// Process the send_join response. The idea here is that we'll try and wait
+	// for as long as possible for the work to complete by using a background
+	// context instead of the provided ctx. If the client does give up waiting,
+	// we'll still continue to process the join anyway so that we don't waste the effort.
+	// TODO: Can we expand Check here to return a list of missing auth
+	// events rather than failing one at a time?
+	respState, err = CheckSendJoinResponse(
+		context.Background(),
+		roomVersion, StateResponse(respSendJoin),
+		input.KeyRing,
+		event,
+		input.EventProvider,
+	)
+	if err != nil {
+		return nil, &FederationError{
+			ServerName: input.ServerName,
+			Transient:  false,
```

Comment:
Hard to say if this is transient as it may have failed due to failing to get server keys over the network.